lib/src/specs/expression.dart

```diff
@@ -0,0 +1,229 @@
+
+/// Represents a Dart expression.
+///
+/// See various concrete implementations for details.
+abstract class Expression implements Spec {
+  const Expression();
+
+  @override
+  R accept<R>(covariant ExpressionVisitor<R> visitor, [R context]);
+
+  /// Returns the expression as a valid [Code] block.
+  Code asCode() => new _AsExpressionCode(this);
+
+  /// Returns the result of [this] `&&` [other].
+  Expression and(Expression other) {
+    return new BinaryExpression._(toExpression(), other, '&&');
+  }
+
+  /// Call this expression as a method.
+  Expression call(
+    List<Expression> positionalArguments, [
+    Map<String, Expression> namedArguments = const {},
+  ]) {
+    return new InvokeExpression._(
+      this,
+      positionalArguments,
+      namedArguments,
+    );
+  }
+
+  /// Returns an expression accessing `.<name>` on this expression.
+  Expression property(String name) {
+    return new BinaryExpression._(
+      this,
+      new LiteralExpression._(name),
+      '.',
+      false,
+    );
+  }
+
+  /// Returns a new instance of this expression.
+  Expression newInstance(
+    List<Expression> positionalArguments, [
+    Map<String, Expression> namedArguments = const {},
+  ]) {
+    return new InvokeExpression._new(
+      this,
+      positionalArguments,
+      namedArguments,
+    );
+  }
+
+  /// Returns a const instance of this expression.
+  Expression constInstance(
+    List<Expression> positionalArguments, [
+    Map<String, Expression> namedArguments = const {},
+  ]) {
+    return new InvokeExpression._const(
+      this,
+      positionalArguments,
+      namedArguments,
+    );
+  }
+
+  /// May be overriden to support other types implementing [Expression].
+  @visibleForOverriding
+  Expression toExpression() => this;
+}
+
+/// Represents a [code] block that wraps an [Expression].
+class _AsExpressionCode implements Code {
+  final Expression code;
+
+  const _AsExpressionCode(this.code);
+
+  @override
+  R accept<R>(CodeVisitor<R> visitor, [R context]) {
+    return code.accept(visitor as ExpressionVisitor<R>, context);
+  }
+}
+
+/// Knowledge of different types of expressions in Dart.
+///
+/// **INTERNAL ONLY**.
+abstract class ExpressionVisitor<T> implements SpecVisitor<T> {
+  T visitBinaryExpression(BinaryExpression expression, [T context]);
+  T visitCodeExpression(CodeExpression expression, [T context]);
+  T visitInvokeExpression(InvokeExpression expression, [T context]);
+  T visitLiteralExpression(LiteralExpression expression, [T context]);
+  T visitLiteralListExpression(LiteralListExpression expression, [T context]);
+  T visitLiteralMapExpression(LiteralMapExpression expression, [T context]);
+}
+
+/// Knowledge of how to write valid Dart code from [ExpressionVisitor].
+///
+/// **INTERNAL ONLY**.
+abstract class ExpressionEmitter implements ExpressionVisitor<StringSink> {
+  @override
+  visitBinaryExpression(BinaryExpression expression, [StringSink output]) {
```

Comment:
Return values?

Reply:
It's automatically inherited in strong-mode semantics (hence why travis is green).

I don't mind explicitly adding, but doesn't add any obvious value in the VM.